LFS-81 / LFS-83: Use the material-dashboard-react to theme the website

OrdiNeu opened the pull request:
Left the code in the Homepage module, since it's going to eventually replace it

marta- committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Added media folder and possible background image

marta- committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Added dependencies to semantic-ui and material-react-dashboard

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Copy material-react-dashboard Dashboard & navbar

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Add sidebar from material-dashboard-react

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Adding logo to Themepage

OrdiNeu committed:
LFS-81 / LFS-83: Pull user data and display on main website
Move Dashboard contents from hard-coded to state-driven

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Remove dead html code that affected page formatting, and alter styles
for aesthetics.

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Remove obsolete files

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Replace the default version of material-ui on the rest of the website
with a downgraded version (3.9.2) as we cannot use the latest version
with material-dashboard-react

OrdiNeu committed:
LFS-71: Replace dashboard icon with sign out icon

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Move magic number into variable and freeze the version of material-dashboard-react to prevent
crashes

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Adding in User Profile route for Bruce

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Clean up dead code

OrdiNeu committed:
LFS-71: Redirect the user to the login page if they are not logged in

OrdiNeu committed:
LFS-71: Replace the default web page

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Fix double scroll bar on themepage

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Remove old /themepage.html link, fix /content links

OrdiNeu committed:
LFS-71: Poll content/forms/ for data

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Prettify the login page

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Remove RTL support

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Further code cleanup

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Close mini-navbar when the window changes width

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Code documentation and cleanup

sdumitriu commented on modules/homepage/pom.xml:
```diff
@@ -45,7 +45,7 @@
         <configuration>
           <instructions>
-            <Include-Resource>src/main/frontend/dist/,{maven-resources}</Include-Resource>
+            <Include-Resource>src/main/frontend/dist/,{maven-resources},media/</Include-Resource>
```
They are still sources, so they should be under `src/main/media`.

sdumitriu commented on modules/homepage/src/main/frontend/src/themePage/Dashboard/dashboard.jsx:
```diff
@@ -0,0 +1,82 @@
+
+// Convert a date into the given format string
+// If the date is invalid (usually because it is missing), return ""
+function _parseDate(date, formatString){
```
"Parse" usually means "convert from a string into another kind of object", so I think this should be called `_formatDate`.

sdumitriu commented on modules/homepage/src/main/frontend/src/themePage/Navbars/AdminNavbarLinks.jsx:
```diff
@@ -0,0 +1,189 @@
+import classNames from "classnames";
+// @material-ui/core components
+import withStyles from "@material-ui/core/styles/withStyles";
+import MenuItem from "@material-ui/core/MenuItem";
```
When not using the `import {Some, Things} from "@material-ui/code"` syntax, the transpiled file will embed MaterialUI, resulting in a fat file.

OrdiNeu replied:
[This isn't possible with @material-ui/core 3.x](https://material-ui.com/blog/material-ui-v4-is-out/#performance), which is the only version compatible with material-dashboard-react. Once I finish with LFS-96 where we remove the material-dashboard-react dependency, we can add tree shaking and cut down on filesize.

sdumitriu replied:
I don't think that's true. What is new isn't the import syntax, but "tree shaking", which affects the size of the webpacked output.

Before 4.0, using `import {This, That} from materialui` resulted in the entire MaterialUI library being included in the output, which, for those not using the entire library and that care about the size of the output, was sub-optimal. This is why `import This; import That` was preferred. However, since we declared MaterialUI as an external library, the output will not include any of MaterialUI, no matter how it is imported. Version 4.0 adds support for tree shaking, or "unused code elimination", which means that starting with MaterialUI 4, both ways of importing will result in the same output size.

sdumitriu replied:
You can analyze the output using [the webpack visualizer](https://chrisbateman.github.io/webpack-visualizer/) to see if it is embedded or not. Then you can do two things:

- you can leave it like that but then it doesn't matter which version of MaterialUI you include in the HTML, the embedded one will be used, so you can keep using the commons header.
- or you can change to the other import syntax, which will give you a smaller file, but you need to make sure the right version of MaterialUI is in the HTML

Also note that they changed how MaterialUI can be referenced from the window, see [this commit](https://github.com/ccmbioinfo/lfs/commit/c4bf80b).

OrdiNeu replied:
You're right; I had figured it didn't work because when I tried it with Navbars/AdminNavbarLinks.jsx it failed -- turns out that there's a typo in the original source code for the Popper component, which caused the `import {this} from '@material-ui/core` syntax to fail.

sdumitriu commented on modules/homepage/src/main/resources/SLING-INF/nodetypes/homepage.cnd:
```diff
@@ -40,3 +40,19 @@
+  // Hardcode the resource supertype: the Homepage is a resource
+  - sling:resourceSuperType (STRING) = "lfs/Resource" mandatory autocreated protected
+  // Set a default title
+  - title (String) = "Welcome to LFS" mandatory autocreated
```
Please always put newlines at the end of files. Otherwise, every time you add something at the end of the file, git will show the last line removed then added right back, messing up the diff.

sdumitriu commented on modules/homepage/src/main/frontend/src/themePage/Sidebar/sidebar.jsx:
```diff
@@ -0,0 +1,142 @@
+  var brand = (
+    <div className={classes.logo}>
+      <a
+        href="https://phenotips.org/"
```
Wrong URL, just use `/`.

sdumitriu commented on modules/homepage/src/main/frontend/src/themePage/routes.jsx:
```diff
@@ -0,0 +1,62 @@
+
+var sidebarRoutes = [
+    {
+      path: "/content",
```
Can you make it work with multiple URLs? This should work for:

- /
- /content
- /content.html

OrdiNeu replied:
None of the answers in [this StackOverflow answer](https://stackoverflow.com/questions/40541994/multiple-path-names-for-a-same-component-in-react-router) seem to work with the version of React-router that material-dashboard-react is using. I can, however, make it work in LFS-96.

OrdiNeu replied:
(For the record, `npx yarn why react-router` gives me two installations of react-router -- v5.0.1 from our version of react-router-dom, and v4.3.1 from material-dashboard-react.)

sdumitriu commented on modules/login/src/main/resources/SLING-INF/content/libs/lfs/login/html.html:
```diff
@@ -16,7 +16,24 @@
   under the License.
 */-->
-<sly data-sly-resource="${@selectors='header'}"/>
+<html>
```
Why?

OrdiNeu replied:
The headers specify material-ui 4.x, which isn't compatible with material-dashboard-react. Once LFS-96 is complete and the material-dashboard-react dependency is removed, I'll switch back to <sly data-sly-resource...>

sdumitriu commented on modules/homepage/src/main/resources/SLING-INF/nodetypes/homepage.cnd:
```diff
@@ -40,3 +40,19 @@
+
+//-----------------------------------------------------------------------------
+// Simple nodetype for identifying the page with the theme.
+[lfs:Themepage] > sling:Folder
```
Is `lfs:Homepage` still in use? If not, can it be removed? Or better, can we just keep using `Homepage` instead?

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Move homepage /media to /src/main/media

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Rename _parseDate to _formateDate

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Replace lfs:Homepage with lfs:Themepage

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Fix sidebar home redirect URL

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Add newlines at the end of files

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Swap to using curly-brace named imports

sdumitriu commented on modules/homepage/src/main/frontend/webpack.config.js:
```diff
@@ -3,7 +3,8 @@ module.exports = {
+    themeindex: './src/themePage/index.jsx'
   },
   module: {
     rules: [
```
I think you need to change the externals config: `"@material-ui/core": "MaterialUI"` becomes `"@material-ui/core": "window['material-ui']"`.

sdumitriu commented on modules/homepage/src/main/frontend/webpack.config.js:
```diff
@@ -3,7 +3,8 @@ module.exports = {
     navbar: './src/navbar/navbarMain.js',
     userSplash: './src/userSplash/userSplashPage.js',
-    index: './src/homePage/index.js'
+    index: './src/homePage/index.js',
```
Is this still needed? Can we clean up the old homepage code, maybe in a follow-up pull request?

OrdiNeu replied:
I can remove it in this PR, since it's in-scope.

sdumitriu commented on modules/login/src/main/resources/SLING-INF/content/libs/lfs/login/html.html:
```diff
@@ -16,7 +16,24 @@
+    <script crossorigin src="https://unpkg.com/jss@9.8.7/dist/jss.js"></script>
+    <script crossorigin src="https://unpkg.com/react@16.8.4/umd/react.development.js"></script>
+    <script crossorigin src="https://unpkg.com/react-dom@16.8.4/umd/react-dom.development.js"></script>
+    <script crossorigin src="https://unpkg.com/@material-ui/core@4.1.1/umd/material-ui.development.js"></script>
```
But you're still using 4.1.1?

OrdiNeu replied:
Wait a sec, this is the login module -- the only reason I removed the header/footer was to get rid of the top bar / bottom text, on Marta's recommendation.

sdumitriu commented on modules/homepage/src/main/frontend/src/themePage/index.jsx:
```diff
@@ -0,0 +1,125 @@
+    return (
+      <div className={classes.wrapper}>
+        <Sidebar
+          routes={sidebarRoutes}
```
I think the reason why multiple URLs for the same thing doesn't work isn't the react router version, but the `Sidebar` doesn't support anything other than a simple string.

sdumitriu replied:
Since we're adopting the code, we could patch it to work with multiple URLs, but in the next sprint.

OrdiNeu replied:
I made sure to incorporate arrays (so that both Sidebar and Navbar used path[0]) but it still behaved very strangely.

sdumitriu commented on modules/homepage/src/main/frontend/src/themePage/routes.jsx:
```diff
@@ -0,0 +1,62 @@
+      component: DashboardPage
+    },
+    {
+      path: "/view.html",
```
To make reloads work, can you try using URLs like `/content.html/view.html`, `/content.html/modelorganisms.html`...

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Move routes from '/<page>' to '/content.html/<page>' to make the
resultant URLs valid, enabling page refreshes

sdumitriu commented on modules/homepage/src/main/frontend/src/themePage/index.jsx:
```diff
@@ -0,0 +1,125 @@
+  <Router history={hist}>
+    <Switch>
+      <Route path="/" component={MainComponent} />
+      <Redirect from="/" to="/content/dashboard.html" />
```
Do you mean this instead?

```suggestion
      <Redirect from="/" to="/content.html/dashboard.html" />
```

OrdiNeu replied:
For some reason, I can't redirect from "/content.html" to anywhere, meaning that visiting /content.html leads you to a blank page with the sidebar.

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Fix redirect links

OrdiNeu committed:
LFS-81 / LFS-82: Use the material-dashboard-react to theme the website
Remove old homepage code
